imperial_coldfront_plugin/gpfs_client.py

```diff
@@ -1,9 +1,100 @@
 """Interface for interacting with the GPFS API."""
 
+from collections.abc import Generator
+
 import requests
 from django.conf import settings
-from uplink import Consumer, get, returns
+from uplink import Body, Consumer, get, json, post, response_handler, retry
 from uplink.auth import BasicAuth
+from uplink.retry.backoff import exponential
+from uplink.retry.stop import after_delay
+from uplink.retry.when import RetryPredicate, status_5xx
+
+
+class ErrorWhenProcessingJob(Exception):
+    """Handles errors in asynchronous jobs."""
+
+
+class JobTimeout(Exception):
+    """Raises an exception when a job times out."""
+
+
+class JobRunning(RetryPredicate):
+    """Checks the job status to decide it there should be a retry."""
+
+    def should_retry_after_response(self, response: requests.Response) -> bool:
+        """Check the response to decide if it should retry.
+
+        Args:
+            response: The response to explore.
+
+        Raises:
+            ErrorWhenProcessingJob: if the job status is FAILED.
+
+        Returns:
+            True if the request should be retried, False otherwise.
+        """
+        if not 200 <= response.status_code < 300:
+            return False
+
+        job_data: dict = response.json()["jobs"][0]
+        if job_data["status"] == "FAILED":
+            job_data["result"]["jobId"] = job_data["jobId"]
+            raise ErrorWhenProcessingJob(job_data["result"])
+        elif job_data["status"] == "RUNNING":
+            return True
+
+        return False
+
+
+class TimeoutWithException(after_delay):
+    """Stops the retry with a custom exception when there is too much delay."""
+
+    def __call__(self) -> Generator[bool, float, None]:
+        """Checks if delay is beyond maximum timeout.
+
+        Raises:
+            JobTimeout: if delay is greater than the maximum delay.
+
+        Yields:
+            False, if delay has not reached the maximum.
+        """
+        while True:
+            delay = yield 0
+            if self._max_delay < delay:
+                raise JobTimeout()
+            yield False
```

Comment:
Can we handle all of the exception raising in `check_job_status`? Not sure what happens with `after_delay` if it times out bet we could presumably detect and just raise the exception in `check_job_status`.